### PR TITLE
ConfigurationBinder.Bind on virtual properties duplicates elements

### LINKED ISSUE #70150: ConfigurationBinder.Bind on virtual array property duplicates elements
### Description

Microsoft.Extensions.Configuration.ConfigurationBinder.Bind sets virtual properties multiple times when overridden. This appends a duplicate array for each overridden instance of virtual array properties.

This appears to come from ConfigurationBinder.GetAllProperties, it adds overridden properties to its returned list multiple times.

### Reproduction Steps

```C#
public class C1
{
    public virtual string[] Test { get; set; } = System.Array.Empty<string>();
}

public class C2 : C1
{
    public override string[] Test { get => base.Test; set => base.Test = value; }
}

var configuration = new ConfigurationBuilder()
    .AddInMemoryCollection(new Dictionary<string, string>()
    {
        ["Test:0"] = "1",
    })
    .Build();

var test = new C2();
configuration.Bind(test);
// test.Test has two elements with value of "1"
```

### Expected behavior

I would expect test.Test to have one element "1"

### Actual behavior

test.Test has two elements with value of "1"

### Regression?

_No response_

### Known Workarounds

_No response_

### Configuration

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by pedrobsaila**
Fixes #70150

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 1]
```diff
@@ -757,11 +757,30 @@ private static bool IsArrayCompatibleReadOnlyInterface(Type type)
             Type type)
         {
             var allProperties = new List<PropertyInfo>();
+            HashSet<string>? propertyNames = null;
 
             Type? baseType = type;
             do
             {
-                allProperties.AddRange(baseType!.GetProperties(DeclaredOnlyLookup));
+                PropertyInfo[] baseTypeProperties = baseType!.GetProperties(DeclaredOnlyLookup);
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:765]: Can't we just use something like:

`type.GetProperties(BindingFlags.Instance | BindingFlags.Public)`

?

Comment (pedrobsaila) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:765]: This breaks some unit tests :

```
[xUnit.net 00:01:15.23]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanReadStaticProperty [FAIL]
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanReadStaticProperty [27 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanReadStaticProperty() in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 964
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
[xUnit.net 00:01:15.27]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanReadStaticProperty [FAIL]
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanReadStaticProperty [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: other stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanReadStaticProperty() in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 981
   at System.RuntimeMethodHandle.InvokeMethod(Object target, Void** arguments, Signature sig, Boolean isConstructor)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
[xUnit.net 00:01:15.29]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "InternalProperty") [FAIL]
[xUnit.net 00:01:15.29]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "PrivateProperty") [FAIL]
[xUnit.net 00:01:15.29]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "ProtectedProperty") [FAIL]
[xUnit.net 00:01:15.29]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "ProtectedPrivateSet") [FAIL]
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "InternalProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1113
   at InvokeStub_ConfigurationBinderTests.BindCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "PrivateProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1113
   at InvokeStub_ConfigurationBinderTests.BindCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "ProtectedProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1113
   at InvokeStub_ConfigurationBinderTests.BindCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(property: "ProtectedPrivateSet") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.BindCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1113
   at InvokeStub_ConfigurationBinderTests.BindCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
[xUnit.net 00:01:15.30]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "InternalProperty") [FAIL]
[xUnit.net 00:01:15.30]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "PrivateProperty") [FAIL]
[xUnit.net 00:01:15.30]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "ProtectedProperty") [FAIL]
[xUnit.net 00:01:15.30]     Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "ProtectedPrivateSet") [FAIL]
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "InternalProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1047
   at InvokeStub_ConfigurationBinderTests.GetCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "PrivateProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1047
   at InvokeStub_ConfigurationBinderTests.GetCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "ProtectedProperty") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1047
   at InvokeStub_ConfigurationBinderTests.GetCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
  Failed Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(property: "ProtectedPrivateSet") [< 1 ms]
  Error Message:
   Assert.Equal() Failure
Expected: stuff
Actual:   (null)
  Stack Trace:
     at Microsoft.Extensions.Configuration.Binder.Test.ConfigurationBinderTests.GetCanSetNonPublicWhenSet(String property) in C:\OSS\runtime\src\libraries\Microsoft.Extensions.Configuration.Binder\tests\ConfigurationBinderTests.cs:line 1047
   at InvokeStub_ConfigurationBinderTests.GetCanSetNonPublicWhenSet(Object, Object, IntPtr*)
   at System.Reflection.MethodInvoker.Invoke(Object obj, IntPtr* args, BindingFlags invokeAttr)
```

Comment (pedrobsaila) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs]: Yes it works now. Just to be sure, changing only binding flags does not deduplicate virtual fields, we still need the hashset

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs]: I was not clear enough on my previous comments. What I meant was:

```C#
        private static List<PropertyInfo> GetAllProperties([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type type) =>
            type.GetProperties(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static);
```

Comment (pedrobsaila) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs]: my bad it was obvious :dizzy_face:

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 2]
```diff
@@ -755,19 +755,7 @@ private static bool IsArrayCompatibleReadOnlyInterface(Type type)
         private static List<PropertyInfo> GetAllProperties(
             [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)]
             Type type)
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:757]:
```suggestion
        private static List<PropertyInfo> GetAllProperties([DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.All)] Type type)
```

File: src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs [hunk 3]
```diff
@@ -755,19 +755,7 @@ private static bool IsArrayCompatibleReadOnlyInterface(Type type)
...
-            do
-            {
-                allProperties.AddRange(baseType!.GetProperties(DeclaredOnlyLookup));
-                baseType = baseType.BaseType;
-            }
-            while (baseType != typeof(object));
-
-            return allProperties;
-        }
+        => type.GetProperties(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static).ToList();
```
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:758]: (nit) The indention here is non-typical. I would make this one more level of indentation and either move the `=>` up to the previous line or leave it on this line. (Along with collapsing the signature into one line so the method is easier to read.)

```suggestion
            => type.GetProperties(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static).ToList();
```

Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs:758]:
```suggestion
        => type.GetProperties(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static);
```

There's no reason to copy this array into a list - it is just extra allocation and copying around data unnecessarily.

**Review by eerhardt — APPROVED** [review]
This looks good to me. Thanks @pedrobsaila!

**Comment by eerhardt** [comment]
`runtime (Build Browser wasm Linux Release LibraryTests)` failure is unrelated to this change.

**Pull request merged**